### PR TITLE
[feaLib] Add support for STAT table

### PR DESCRIPTION
**Pull request by punchcutter**
Add STAT table support as discussed in https://github.com/adobe-type-tools/afdko/issues/176

One issue not yet updated in afdko is discussed here: https://github.com/adobe-type-tools/afdko/issues/1197
This change is already reflected in the feaLib implementation.

### CODE REVIEW

File: Tests/feaLib/builder_test.py
Comment (moyogo) [inline Tests/feaLib/builder_test.py:484]: This should assert() the table is correct instead of print().

File: Lib/fontTools/ttLib/tables/otConverters.py
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/otConverters.py:337]: Maybe a generic NamedFlags converter that could be defined in otConverters and used anywhere there are some bit flags that we want to be accompanied by a comment with the flag names. Perhaps add a TODO for later.

File: Lib/fontTools/feaLib/builder.py
Comment (khaledhosny) [inline Lib/fontTools/feaLib/builder.py:533]: We recently got `otlLib.builder.buildStatTable()`, can’t it be used here?

File: Lib/fontTools/feaLib/ast.py
Comment (khaledhosny) [inline Lib/fontTools/feaLib/ast.py:90]: There are many name entries elsewhere in this module (`NameRecord`, `FeatureNameStatement`, `CVParametersNameStatement`) and they don’t do any escaping. I suggest you try to match their code for now, and open a separate issue if you think it should do escaping. It is better to be consistent here, IMO.

File: Lib/fontTools/feaLib/ast.py
Comment (khaledhosny) [inline Lib/fontTools/feaLib/ast.py:1885]: Instead of using a tuple for name record, it might be better to subclass `FeatureNameStatement` and override its `build` method.

File: Lib/fontTools/feaLib/ast.py
Comment (khaledhosny) [inline Lib/fontTools/feaLib/ast.py:1905]: Same as above.

File: Tests/feaLib/builder_test.py
Comment (punchcutter) [reply Tests/feaLib/builder_test.py:484]: Argh, I thought I removed that whole thing. The "test" is in another place.

File: Lib/fontTools/feaLib/builder.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py:1440]: Nothing, but I had started doing some vmtx stuff on the same branch and failed to remove this part.

File: Lib/fontTools/feaLib/builder.py
Comment (moyogo) [reply Lib/fontTools/feaLib/builder.py:533]: It could be used but the logic of `otLib.build.buildStatTable()` is rather different from what FEA syntax does, which is closer to what AxisValueRecords do, and it also doesn’t support languageIDs.
Basically `otLib.build.buildStatTable()` was written to be easy to use, but is not easy to use with FEA syntax.

Comment (anthrotype) [reply Lib/fontTools/feaLib/builder.py:533]: Thanks Khaled for reminding me of that. Yes, we should definitely try to reuse that code in here too, see whether it's possible to modify it to accommodate both.

File: Lib/fontTools/feaLib/ast.py
Comment (anthrotype) [reply Lib/fontTools/feaLib/ast.py:90]: agreed

File: Lib/fontTools/feaLib/ast.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/ast.py:88]: We don't actually need any of this. At first I was trying to make sure tests and round-tripping were working to some degree, but feaLib can't actually do that currently so I dropped all of this.

File: Lib/fontTools/feaLib/ast.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/ast.py:1901]: Tell that to Felipe. He likes to keep the f in fontbakery.

File: Lib/fontTools/feaLib/ast.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/ast.py:1963]: Ok, I prefer single quotes as you can see, but I'm fine doing double to try to be more consistent.

File: Lib/fontTools/feaLib/parser.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/parser.py:1300]: We don't actually need it at all.

File: Lib/fontTools/ttLib/tables/otConverters.py
Comment (punchcutter) [reply Lib/fontTools/ttLib/tables/otConverters.py:345]: Updated all of this to do the same as lookup flags.

File: Lib/fontTools/ttLib/tables/otConverters.py
Comment (punchcutter) [reply Lib/fontTools/ttLib/tables/otConverters.py:337]: Updated to do the same as lookup flags.

File: Lib/fontTools/feaLib/ast.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/ast.py:90]: Done. Dropped all this and used FeatureNameStatement.

File: Lib/fontTools/feaLib/ast.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/ast.py:1885]: Actually, there's no reason to subclass FeatureNameStatement. These can just be FeatureNameStatement objects because they behave the same.

Comment (khaledhosny) [reply Lib/fontTools/feaLib/ast.py:1885]: I was concerned that its `build()` method might not be appropriate for `STAT` table, but if it works then all good.

File: Lib/fontTools/feaLib/parser.py
Comment (khaledhosny) [inline Lib/fontTools/feaLib/parser.py:1345]: I‘m not sure why do you need `expect_stat_values_()` at all, since you are already checking the token type, you can simplify this to `self.advance_lexer_()` and read the value from `self.cur_token_`.

Comment (punchcutter) [reply Lib/fontTools/feaLib/parser.py:1345]: I made it separate originally because we also had the hyphen for ranges and it seemed worth it to move to a separate `expect_`, but you're right, there's no need now.

File: Lib/fontTools/feaLib/ast.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/ast.py]: redid all of this to match the rest of feaLib

File: Lib/fontTools/feaLib/builder.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py]: updated everything (I think)

File: Lib/fontTools/feaLib/ast.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/ast.py]: Actually, you're right. The `build()` method for FeatureNameStatement does some stuff we don't want to do with STAT names so I subclassed NameRecord to get what we need.

File: Tests/feaLib/builder_test.py
Comment (anthrotype) [inline Tests/feaLib/builder_test.py:154]: what are you trying to do here? Why split at semicolons?
Doesn't the `normal_fea` method already return list of lines stripped of the ending newline character?

File: Tests/feaLib/runit.py
Comment (anthrotype) [inline Tests/feaLib/runit.py:59]: what is this `runit.py` script?

File: Lib/fontTools/feaLib/builder.py
Comment (anthrotype) [reply Lib/fontTools/feaLib/builder.py:552]: @punchcutter have you tried to see if there's anything in otlLib that you can reuse in here?
https://github.com/fonttools/fonttools/blob/bec19252c5b424a88e1dfeb423fb2c5068fb36e9/Lib/fontTools/otlLib/builder.py#L2399-L2590

File: Lib/fontTools/ttLib/tables/otConverters.py
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/otConverters.py:56]: remove this `elif` block, add a `STATFlags: STATFlags` to the `converterMapping` below, then change the entry for the STAT Flags in `otData.py` module so that the type of the converter is `STATFlags` instead of `uint16`

File: Tests/feaLib/builder_test.py
Comment (punchcutter) [reply Tests/feaLib/builder_test.py:154]: Like I put in the TODO above this way of comparing files has all kinds of problems so the comparison only works for some things and assumes all lines are exactly the same. Splitting on semicolons makes more sense than newlines because semicolons have meaning, but newlines can be arbitrary. Also, spaces, including newlines, are only relevant within a string so to truly compare fea we should also ignore all spaces except in strings. That's why I only did this minor adjustment and added a TODO. This can be much better, but isn't necessary right now.

File: Tests/feaLib/runit.py
Comment (punchcutter) [reply Tests/feaLib/runit.py:59]: Oops. Not really sure where that came from. Deleted.

File: Lib/fontTools/ttLib/tables/otConverters.py
Comment (punchcutter) [reply Lib/fontTools/ttLib/tables/otConverters.py:56]: Sounds good.

File: Lib/fontTools/feaLib/builder.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py:552]: The base build code was taken from there since it was already there. Of course it's possible to merge the two, but I wanted to get this working in feaLib first before messing with that.

File: Lib/fontTools/feaLib/builder.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py:514]: Ok

Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py]: Actually, makes sense to consolidate these since we can check if the value is an int or list and assign it with that.

File: Lib/fontTools/feaLib/ast.py
Comment (felipesanches) [reply Lib/fontTools/feaLib/ast.py]: @punchcutter that was not exactly my point though. I agree with @anthrotype's suggestion here.

Comment (felipesanches) [reply Lib/fontTools/feaLib/ast.py]: In FontBakery codebase we use the f-string prefix on all lines of a multi-line string for the sake of keeping the code's visual alignment "beautiful" in cases like this:

```python
print(f'This is and example in which'
      f' the value of xyz is {xyz} and'
      f' blah blah blah whatever...')
```

but this is not the case here, where the string in `res += ' };'` is isolated in its own statement. There's no need for visual alignment here, so it is completely unrelated to the code-style adopted in fontbakery

File: Lib/fontTools/feaLib/builder.py
Comment (anthrotype) [inline Lib/fontTools/feaLib/builder.py:514]: even if unlikely it's always worth adding a final else block for self-protection  
```
else:
    raise AssertionError(value)
```

File: Lib/fontTools/feaLib/builder.py
Comment (anthrotype) [inline Lib/fontTools/feaLib/builder.py:552]: in bare-bone fonts (e.g. unit tests) the name table may not be present, here's what we do elsewhere in feaLib.builder

https://github.com/fonttools/fonttools/blob/a41005077dc36f27798e64d659934aa3dcdf016a/Lib/fontTools/feaLib/builder.py#L391-L394

File: Tests/feaLib/builder_test.py
Comment (anthrotype) [reply Tests/feaLib/builder_test.py:154]: > Splitting on semicolons makes more sense than newlines because semicolons have meaning, but newlines can be arbitrary.

but the arbitrary newlines are the ones that feaLib asFea() itself puts, so at least they should be consistent and predictable. 
Anyway I'm not convinced this is necessary for the present PR. I would prefer if this was addressed separately if at all.

**Review by anthrotype — COMMENTED** [review]
Apologies if I asked this already, have you tried to see if we can reuse/import code from the buildStatTable in otlLib?

Since this is quite a lot of code that touches on several places, i'd like someone else (e.g. @justvanrossum @simoncozens @khaledhosny @dscorbett) to also give their LGTM

File: Lib/fontTools/feaLib/builder.py
Comment (anthrotype) [reply Lib/fontTools/feaLib/builder.py:546]: yes I agree

File: Lib/fontTools/feaLib/builder.py
Comment (justvanrossum) [reply Lib/fontTools/feaLib/builder.py:514]: Also, that second if should be an elif.

File: Lib/fontTools/feaLib/builder.py
Comment (justvanrossum) [reply Lib/fontTools/feaLib/builder.py:546]: I also agree, and would be happy to review changes or refactorings to `fontTools.otlLib.builder.buildStatTable()` to make this possible.

Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py:546]: Ok, my initial thought was to get this in ASAP so we'd have it for some existing projects that are already set up with STAT in the feature file. And then after that definitely do a refactoring at some point. Since this is already quite a lot of new stuff I felt like it might be good to leave the otlLib code alone for a first round, but either way is fine.

File: Tests/feaLib/builder_test.py
Comment (punchcutter) [reply Tests/feaLib/builder_test.py:154]: Ok, fine to drop it from this PR.

File: Lib/fontTools/feaLib/builder.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py:552]: Replaced

File: Lib/fontTools/feaLib/builder.py
Comment (punchcutter) [reply Lib/fontTools/feaLib/builder.py:514]: Updated

File: Tests/feaLib/data/STAT_test.fea
Comment (anthrotype) [inline Tests/feaLib/data/STAT_test.fea:13]: why this?

Comment (punchcutter) [reply Tests/feaLib/data/STAT_test.fea:15]: Why what? New line? I just updated now so it's consistent with the rest. In practice people can do whatever they want with the whitespace, but here we expect newlines for each name entry since there can be multiple.

File: Lib/fontTools/feaLib/ast.py
Comment (anthrotype) [inline Lib/fontTools/feaLib/ast.py:70]: you'll notice the convention in feaLib.ast module is to name the classes with an ending "*Statement" when they consist of a "statement" (i.e. a line of FEA ending with semicolon). It may be good to keep that convention for consistency. All the new classes defined here are statements in this sense. Mind renaming them, even though the names becomes a bit long?

File: Lib/fontTools/feaLib/ast.py
Comment (anthrotype) [inline Lib/fontTools/feaLib/ast.py:2013]: any reason why this is a namedtuple and not an ast.Statement like the others?
I think you turn it into a statement object and add asFea() method as well, so STATAxisValueRecord can use that in its asFea

**Review by anthrotype — COMMENTED** [review]
Thank you for your first contribution Kamile!
I added some inline review comments below, please take a look.

After all reviewers have added their LGTM, do you mind also running the modules you touched through the [black](https://github.com/psf/black) auto-formatter, to normalize minor whitespace issues? Not before we're ready to merge though, otherwise we can't distinguish what changed since the last review pass.

File: Lib/fontTools/feaLib/builder.py
Comment (kaydeearts) [reply Lib/fontTools/feaLib/builder.py:576]: `axisValueRecords` may return as None instead of an empty list, so I will specify this to be `if axisValueRecords is not None:`. Without the check, there is a _NoneType Object is not iterable_ error

**Comment by kaydeearts** [comment]
@anthrotype Thank you for your review, I've gone through all of your comments & updated the code based on them.

File: Lib/fontTools/otlLib/builder.py
Comment (anthrotype) [inline Lib/fontTools/otlLib/builder.py:2806]: hm.. here you're assuming that the namerecord is a windows-platform one and you're only using its langID, i don't think you can make this assumption. 
it kind of was more correct before this last edit... You were actually right, we don't want to use `addMultilingualName` when we already have a list of NameRecords.

What we want to do here is basically append the name records, it turns out that the name table has an `addName` method that does exactly what you want: i.e. adds a string with given triplet encoding and assigns it to an unused nameID slot in the user range > 255.
Just use that instead of the combo _findUnusedNameID + setName.

Also, when _addName gets a list, make sure that all the values are instances of `_n_a_m_e.NameRecord` (raise TypeError if they aren't). Document that in the buildStatTable docstring that `elidedFallbackName` can also be a list of namerecords (in addition to a int or a dict).

File: Tests/feaLib/data/STAT_test.ttx
Comment (anthrotype) [inline Tests/feaLib/data/STAT_test.ttx:12]: these duplicated entries are an unintentional byproduct of using addMultilingualName early on (which has an option mac=True), when given a list of elidedFallbackName records. We only want to add the records that we are requested, we don't want to automatically add "mac" names unless explicitly requested. Using the simple addName (as I commented above) should do what we want and remove these duplicate entries.

File: Lib/fontTools/otlLib/builder.py
Comment (kaydeearts) [reply Lib/fontTools/otlLib/builder.py:2806]: @anthrotype I tried using `addName` where it iterates through `value` and is called for each name. However, I realized it gives a new nameID for each name every time, which creates a problem for an ElidedFallBackName with multiple names. For example:

`table STAT {
    ElidedFallbackName {
        name "Roman";
        name 3 1 1041 "ローマン";
    };`

I was going to suggest this, which still uses `addMultilingualName` but specifies which platform is being requested, and keeps the nameID the same for multiple names in ElidedFallbackName:

`        names = {}
        mac = False
        windows = False
        for name in value:
            if name.platformID == 1:
                langCode = _MAC_LANGUAGES.get(name.langID)
                mac = True
            else:
                langCode = _WINDOWS_LANGUAGES.get(name.langID)
                windows = True
            names[langCode] = name.string
        return nameTable.addMultilingualName(names, windows=windows, mac=mac, minNameID=minNameID)`

Comment (anthrotype) [reply Lib/fontTools/otlLib/builder.py]: I see. Ok, then I suggest we revert to using setName like you were doing originally. It's probably fine that you use `_findUnusedNameID` even if underscore-prefixed and intended to be private.

**Comment by anthrotype** [comment]
Thanks, I think this is looking good now. Maybe @justvanrossum would like to take another pass before we run black autoformatter, squash the commits and merge?

**Review by justvanrossum — APPROVED** [review]
I didn't manage a proper review, but overall this looks good to me.

**Comment by anthrotype** [comment]
Thanks Just for the quick review.

I just run the `black` command on the feaLib modules that were already blackened in https://github.com/fonttools/fonttools/pull/2039/commits/6b87f5588c9e9cdfce9125e6bead91ba55b38476.

I am going to merge now. Thanks again to @punchcutter who initiated this PR as well as @kaydeearts for completing it! 👍

**Pull request merged**
